File: objc/Project2/Project2Tests/NotificationTest.m
//
//  NotificationTest.m
//  Project2Tests
//
//  Created by Jinwoo Kim on 2/24/21.
//

#import <XCTest/XCTest.h>

static NSNotificationName upgradedNotification = @"UserUpgraded";

@interface User : NSObject
- (void)upgrade;
- (void)upgrade2;
- (void)upgrade3Using:(NSNotificationCenter *)center;
@end

@implementation User
- (void)upgrade {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [NSThread sleepForTimeInterval:1];
        NSNotificationCenter *center = NSNotificationCenter.defaultCenter;
        [center postNotificationName:upgradedNotification object:nil];
    });
}

- (void)upgrade2 {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [NSThread sleepForTimeInterval:1];
        NSNotificationCenter *center = NSNotificationCenter.defaultCenter;
        [center postNotificationName:upgradedNotification object:nil userInfo:@{@"level": @"gold"}];
    });
}

// default는 Singleton이므로 충돌을 방지하기 위해 Test에서는 쓰지 말자.
- (void)upgrade3Using:(NSNotificationCenter *)center {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [NSThread sleepForTimeInterval:1];
        [center postNotificationName:upgradedNotification object:nil userInfo:@{@"level": @"gold"}];
    });
}
@end

@interface NotificationTest : XCTestCase
@end

@implementation NotificationTest
// 3초동안 Notification을 기다려보고 안 오면 Fail
- (void)testUserUpgradedPostsNotification {
    // given
    User *user = [User new];
    // notificationCenter에 대한 정의가 없으면 default를 옵저빙한다.
    XCTNSNotificationExpectation *expectation = [[XCTNSNotificationExpectation alloc] initWithName:upgradedNotification];
    
    // when
    [user upgrade];
    
    // then
    [self waitForExpectations:@[expectation] timeout:3];
}

- (void)testUserUpgradedPostsNotification2 {
    // given
    User *user = [User new];
    // notificationCenter에 대한 정의가 없으면 default를 옵저빙한다.
    XCTNSNotificationExpectation *expectation = [[XCTNSNotificationExpectation alloc] initWithName:upgradedNotification];
    
    // 들어온 Notification에 따라 NO이면 Fail, YES이면 Pass
    [expectation setHandler:^BOOL(NSNotification * _Nonnull notification) {
        NSString *level = notification.userInfo[@"level"];
        if (level == nil) return NO;
        
        if ([level isEqualToString:@"gold"]) {
            return YES;
        } else {
            return NO;
        }
    }];
    
    // when
    [user upgrade2];
    
    // then
    [self waitForExpectations:@[expectation] timeout:3];
}

- (void)testUserUpgradedPostsNotification3 {
    // given
    NSNotificationCenter *center = [NSNotificationCenter new];
    User *user = [User new];
    XCTNSNotificationExpectation *expectation = [[XCTNSNotificationExpectation alloc] initWithName:upgradedNotification object:nil notificationCenter:center];
    
    // 들어온 Notification에 따라 NO이면 Fail, YES이면 Pass
    [expectation setHandler:^BOOL(NSNotification * _Nonnull notification) {
        NSString *level = notification.userInfo[@"level"];
        if (level == nil) return NO;
        
        if ([level isEqualToString:@"gold"]) {
            return YES;
        } else {
            return NO;
        }
    }];
    
    // when
    [user upgrade3Using:center];
    
    // then
    [self waitForExpectations:@[expectation] timeout:3];
}

@end

File: swift/Project2/Project2Tests/NotificationTest.swift
//
//  NotificationTest.swift
//  Project2Tests
//
//  Created by Jinwoo Kim on 2/24/21.
//

import XCTest

struct User {
    static let upgradedNotification = Notification.Name("UserUpgraded")
    
    func upgrade() {
        DispatchQueue.global().async {
            Thread.sleep(forTimeInterval: 1)
            let center = NotificationCenter.default
            center.post(name: User.upgradedNotification, object: nil)
        }
    }
    
    func upgrade2() {
        DispatchQueue.global().async {
            Thread.sleep(forTimeInterval: 1)
            let center = NotificationCenter.default
            center.post(name: User.upgradedNotification, object: nil, userInfo: ["level": "gold"])
        }
    }
    
    // default는 Singleton이므로 충돌을 방지하기 위해 Test에서는 쓰지 말자.
    func upgrade3(using center: NotificationCenter = NotificationCenter.default) {
        DispatchQueue.global().async {
            Thread.sleep(forTimeInterval: 1)
            center.post(name: User.upgradedNotification, object: nil, userInfo: ["level": "gold"])
        }
    }
}

class NotificationTest: XCTestCase {
    // 3초동안 Notification 알림을 기다려보고 안 오면 Fail
    func testUserUpgradedPostsNotification() {
        // given
        let user = User()
        // notificationCenter에 대한 정의가 없으면 default를 옵저빙한다.
        let expectation = XCTNSNotificationExpectation(name: User.upgradedNotification)
        
        // when
        user.upgrade()
        
        // then
        wait(for: [expectation], timeout: 3)
    }
    
    func testUserUpgradedPostsNotification2() {
        // given
        let user = User()
        // notificationCenter에 대한 정의가 없으면 default를 옵저빙한다.
        let expectation = XCTNSNotificationExpectation(name: User.upgradedNotification)
        
        // 들어온 Notification에 따라 false이면 Fail, true이면 Pass
        expectation.handler = { notification -> Bool in
            guard let level = notification.userInfo?["level"] as? String else {
                return false
            }
            
            if level == "gold" {
                return true
            } else {
                return false
            }
        }
        
        // when
        user.upgrade2()
        
        // then
        wait(for: [expectation], timeout: 3)
    }
    
    func testUserUpgradePostsNotification3() {
        // given
        let center = NotificationCenter()
        let user = User()
        let expectation = XCTNSNotificationExpectation(name: User.upgradedNotification, object: nil, notificationCenter: center)
        
        expectation.handler = { notification -> Bool in
            guard let level = notification.userInfo?["level"] as? String else {
                return false
            }
            
            if level == "gold" {
                return true
            } else {
                return false
            }
        }
        
        // when
        user.upgrade3(using: center)
        
        // then
        wait(for: [expectation], timeout: 3)
    }
}
